book/src/dev/rfcs/0005-state-updates.md

```diff
@@ -0,0 +1,697 @@
+**Note**: Unlike `zcashd`, Zebra does not use block arrival times as a tie-breaker for the best tip. Since Zebra downloads blocks in parallel, download times are not guaranteed to be unique. Using the `block::Hash` provides a consistent tip order. (As a side-effect, the tip order is also consistent after a node restart, and between nodes.)
+### `ChainSet` Type
+[chainset-type]: #chainset-type
+
+The `ChainSet` type represents the set of all non-finalized state. It
+consists of a set of non-finalized but verified chains and a set of
+unverified blocks which are waiting for the full context needed to verify
+them to become available.
+
+`ChainSet` is defined by the following structure and API:
+
+```rust
+struct ChainSet {
+    chains: BTreeSet<Chain>,
+
+    queued_blocks: BTreeMap<block::Hash, QueuedBlock>,
+    queued_by_parent: BTreeMap<block::Hash, Vec<block::Hash>>,
+    queued_by_height: BTreeMap<block::Height, Vec<block::Hash>>,
+}
+```
+
+#### `pub fn finalize(&mut self) -> Arc<Block>`
+
+Finalize the lowest height block in the non-finalized portion of the best
+chain and updates all side chains to match.
+
+1. Extract the best chain from `self.chains` into `best_chain`
+
+2. Extract the rest of the chains into a `side_chains` temporary variable, so
+   they can be mutated
+
+3. Remove the lowest height block from the best chain with
+   `let block = best_chain.pop_root();`
+
+4. Add `best_chain` back to `self.chains`
+
+5. For each remaining `chain` in `side_chains`
+    - If `chain` starts with `block`, remove `block` and add `chain` back to
+    `self.chains`
+    - Else, drop `chain`
+
+6. for each `height` in `self.queued_by_height` where the height is lower than the
+   new reorg limit
+   - for each `hash` in `self.queued_by_height.remove(height)`
+     - Remove the key `hash` from `self.queued_blocks` and store the removed `block`
+     - Find and remove `hash` from `self.queued_by_parent` using `block.parent`'s hash
+
+7. Return `block`
+
+### `pub fn queue(&mut self, block: QueuedBlock)`
+
+Queue a non-finalized block to be committed to the state.
+
+After queueing a non-finalized block, this method checks whether the newly
+queued block (and any of its descendants) can be committed to the state
+
+1. Check if the parent block exists in any current chain
+
+2. If it does, call `let ret = self.commit_block(block)`
+    - Call `self.process_queued(new_parents)` if `ret` is `Some`
+
+3. Else Add `block` to `self.queued_blocks` and related members and return
+
+### `fn process_queued(&mut self, new_parent: block::Hash)`
+
+1. Create a list of `new_parents` and populate it with `new_parent`
+
+2. While let Some(parent) = new_parents.pop()
+    - for each `hash` in `self.queued_by_parent.remove(&parent.hash)`
+      - lookup the `block` for `hash`
+      - remove `block` from `self.queued_blocks`
+      - remove `hash` from `self.queued_by_height`
+      - let result = `self.commit_block(block)`;
+      - add `result` to `new_parents`
```

Comment:
The borrow-checker generally doesn't like loops that modify the arrays they are iterating through - but this is fine for a spec

Reply:
That is only an issue if you're borrowing from the container you're iterating over I think. In this case both operations remove items from the lists before iterating over them so there shouldn't be any issues with the borrow checker.